### PR TITLE
[Reviewer: Rob] Validate json files in the upload_json script, and log when they fail

### LINKED ISSUE #1609: upload_enum_json (and presumably other JSON uploaders) make it easy to upload an invalid file
#### Symptoms

I accidentally created an enum.json that was invalid JSON, and ran `upload_enum_json`, and it accepted it without a problem. (I could diagnose it after the fact from a "The ENUM file is invalid" ENT log.)

It feels like it would be fairly straightforward to automatically check that a JSON file is valid before uploading it, e.g. by running:

`python -c "import json; json.load(open('/etc/clearwater/enum.json'))" 2&>1`

and checking `$?`.

#### Impact

If all you do is `upload_enum_json`, I think the only impact will be confusion about why the new rules aren't taking effect (as the old rules don't get overwritten).

If you make a shared_config change at the same time, however (which the docs don't warn against):

- the `upload_shared_config` script will cause Sprout to restart
- it won't be able to read the s-cscf.json/enum.json/bgcf.json file because it's invalid
- it will start up without any rules, but will otherwise start up normally, so the queue manager will restart all the nodes in the cluster
- if the ENUM, S-CSCF or BGCF rules are necessary for calls to succeed, there might be an outage

#### Release and environment 

`1.0-cc-10.0.0-20161116-~255.00.0-161116.175811` on VMware.

#### Steps to reproduce

Create an enum.jsonfile that looks like `{"number_blocks": []` and run `sudo /usr/share/clearwater/clearwater-config-manager/scripts/
upload_enum_json`.

### PR DESCRIPTION
**Pull request by johadalin**
Rob, 
We thought we had fixed #1609 , but had actually not done so amid a bit of confusion.

I've put the fix you suggested in now. Can you take a quick look at this to assess:

* Is putting this validation in the base `upload_json` file the right thing. i.e. can you think of anything we upload that is invalid, but still uses the upload_json script?
* Are you happy that this can go in at this point? We thought it was in before, and i think it's no-risk, but want to check. Andrew was going to test it anyway.

### CODE REVIEW

**Comment by johadalin** [comment]
Ok. Fixed this up now. I've properly live tested, and confirmed valid json passes, invalid fails with the log, and is not uploaded.

File: sprout-base.root/usr/share/clearwater/clearwater-config-manager/scripts/upload_json [hunk 1]
```diff
@@ -80,6 +80,14 @@ then
   exit 2
 fi
 
+# Ensure that the file is valid JSON
+python -c "import json; json.load(open(\"$CONFPATH/$FILENAME\"))" 2&>1
```
Comment (rkd-msw) [inline sprout-base.root/usr/share/clearwater/clearwater-config-manager/scripts/upload_json]: `python -m json.tool < $CONFPATH/$FILENAME > /dev/null 2>&1` might be a bit cleaner (as you avoid inline Python and escaping bugs).

File: sprout-base.root/usr/share/clearwater/clearwater-config-manager/scripts/upload_json [hunk 2]
```diff
@@ -80,6 +80,14 @@ then
...
+rc=$?
+if [ $rc != 0 ]; then
+  echo "File $CONFPATH/$FILENAME contains invalid json, and so will not be uploaded"
```
Comment (rkd-msw) [inline sprout-base.root/usr/share/clearwater/clearwater-config-manager/scripts/upload_json]: This script now has five different error messages:

```
The Clearwater Configuration store (etcd) is not running
Please start it before uploading configuration
```

```
No configuration file at $CONFPATH/$FILENAME, unable to upload
```

```
File is larger than the recommended size 10000 bytes, use --allow-large if you want to upload
```

```
File $CONFPATH/$FILENAME contains invalid json, and so will not be uploaded
```

```
Upload JSON configuration file failed to $keypath
```

They all just feel inconsistent - is it "JSON" or "json"? Is it a "file" or a "configuration file"? Do we say "please" when asking the user to do something? Do we say "unable to upload" or "will not be uploaded" (or not explicitly say that)?

Should we have and follow a style guide for this help text, along the lines of https://www.postgresql.org/docs/9.5/static/error-style-guide.html or https://www.gov.uk/service-manual/design/help-text?

**Comment by rkd-msw** [comment]
Let's just put it into V11 at this point (once you've addressed my markups), not V10.

File: sprout-base.root/usr/share/clearwater/clearwater-config-manager/scripts/upload_json (hunk 1 above)
Comment (rkd-msw) [inline sprout-base.root/usr/share/clearwater/clearwater-config-manager/scripts/upload_json]: Neither option (the current thing or my suggestion) will upload an empty file. Do we need to handle that?

**Comment by steelyvoid** [comment]
Adding dummy comment to get notifications

**Comment by johadalin** [comment]
closing this PR, as i've rebased against master

**Pull request closed without merging**